src/wovnio/html/HtmlConverter.php

```diff
@@ -38,18 +39,88 @@ public function __construct($html, $encoding, $token, $store, $headers)
     $this->headers = $headers;
   }
 
+  public function insertSnippetAndHreflangTags() {
+    $this->html = $this->insertSnippet($this->html);
+    $this->html = $this->insertHreflangTags($this->html);
+    $marker = new HtmlReplaceMarker();
+    return array($this->html, $marker);
+  }
+
   /**
    * Convert to appropriate HTML to send to Translation API
    * e.g.) remove wovn-ignore content
    *
    * @return array converted html and HtmlReplaceMarker
    */
-  public function insertSnippetAndHreflangTags()
+  public function convertToAppropriateForApiBody()
   {
-    $this->html = $this->insertSnippet($this->html);
-    $this->html = $this->insertHreflangTags($this->html);
+    if ($this->encoding && in_array($this->encoding, self::$supported_encodings)) {
+      $encoding = $this->encoding;
+    } else {
+      // Encoding detection uses 30% of execution time for this method.
+      $encoding = mb_detect_encoding($this->html, self::$supported_encodings);
+    }
+
+    $dom = simple_html_dom::str_get_html($this->html, $encoding, false, false, $encoding, false);
+
+    $marker = new HtmlReplaceMarker();
+    $this->replaceDom($dom, $marker);
 
-    return array($this->html);
+    $converted_html = $dom->save();
+    $converted_html = $this->removeBackendWovnIgnoreComment($converted_html, $marker);
+
+    // Without clear(), Segmentation fault will be raised.
+    // @see https://sourceforge.net/p/simplehtmldom/bugs/103/
+    $dom->clear();
+    unset($dom);
+    return array($converted_html, $marker);
+  }
+
+  private function replaceDom($dom, &$marker) {
+    $self = $this;
+    $adds_hreflang = isset($this->store) && isset($this->headers);
+
+    $html = null;
+    $head = null;
+    $body = null;
+
+    $dom->iterateAll(function ($node) use (&$self, $marker, $adds_hreflang, &$html, &$head, &$body) {
+      if (strtolower($node->tag) == "html") {
+        $html = $node;
+      } else if (strtolower($node->tag) == "head") {
+        $head = $node;
+      } else if (strtolower($node->tag) == "body") {
+        $body = $node;
+      }
+      $self->_removeSnippet($node);
+      if ($adds_hreflang) {
+        $self->_removeHreflang($node);
+      }
+      $self->_removeWovnIgnore($node, $marker);
+      $self->_removeForm($node, $marker);
+      // inside <script>, comment("<!--") is invalid
+      $self->_removeScript($node, $marker);
+    });
+
+    $tags = array($head, $body, $html);
+    foreach ($tags as $insert_tag) {
+      if (is_null($insert_tag)) {
+        continue;
+      }
+
+      $hreflangTags = array();
+      if ($adds_hreflang) {
+        $lang_codes = $this->store->settings['supported_langs'];
+        foreach ($lang_codes as $lang_code) {
+          $href = $this->buildHrefLang($lang_code);
+          array_push($hreflangTags, '<link rel="alternate" hreflang="' . Lang::iso639_1Normalization($lang_code) . '" href="' . $href . '">');
+        }
+      }
+
+      $snippet = $this->buildSnippetCode();
```

Comment:
I think we should pass parameter to `buildSnippetCode` such as `$this->buildSnippetCode($without_backend = false)` to determine whether adding `data-wovnio-type="backend_without_api"` or not

Reply:
thanks, done